File: typescript/bc-envelope/src/functions-store.ts
import {
    type EnvelopeFunction,
    functionName,
    knownFunction,
} from './function.js';

export class FunctionsStore {
    readonly #dict = new Map<string, string>();

    constructor(functions: Iterable<EnvelopeFunction> = []) {
        for (const value of functions) {
            this.insert(value);
        }
    }

    #key(value: EnvelopeFunction): string {
        return value.kind === 'known' ? `k:${value.value}` : `n:${value.name}`;
    }

    insert(value: EnvelopeFunction): void {
        if (value.kind !== 'known') {
            throw new Error('Only known functions can be inserted');
        }
        this.#dict.set(this.#key(value), functionName(value));
    }

    assignedName(value: EnvelopeFunction): string | undefined {
        return this.#dict.get(this.#key(value));
    }

    name(value: EnvelopeFunction): string {
        return this.assignedName(value) ?? functionName(value);
    }

    static nameForFunction(value: EnvelopeFunction, store?: FunctionsStore): string {
        return store?.assignedName(value) ?? functionName(value);
    }
}

export const ADD = knownFunction(1n, 'add');
export const SUB = knownFunction(2n, 'sub');
export const MUL = knownFunction(3n, 'mul');
export const DIV = knownFunction(4n, 'div');
export const NEG = knownFunction(5n, 'neg');
export const LT = knownFunction(6n, 'lt');
export const LE = knownFunction(7n, 'le');
export const GT = knownFunction(8n, 'gt');
export const GE = knownFunction(9n, 'ge');
export const EQ = knownFunction(10n, 'eq');
export const NE = knownFunction(11n, 'ne');
export const AND = knownFunction(12n, 'and');
export const OR = knownFunction(13n, 'or');
export const XOR = knownFunction(14n, 'xor');
export const NOT = knownFunction(15n, 'not');

export const GLOBAL_FUNCTIONS = new FunctionsStore([ADD, SUB, MUL, DIV]);

File: go/bcenvelope/functions_store.go
package bcenvelope

import "sync"

// FunctionsStore maps functions to their assigned names.
type FunctionsStore struct {
	dict map[uint64]functionEntry   // known functions keyed by value
	named map[string]functionEntry  // named functions keyed by name string
}

type functionEntry struct {
	function Function
	name     string
}

// NewFunctionsStore creates a new store populated with the provided functions.
// Only known (numeric) functions can be inserted.
func NewFunctionsStore(functions ...Function) *FunctionsStore {
	store := &FunctionsStore{
		dict:  make(map[uint64]functionEntry),
		named: make(map[string]functionEntry),
	}
	for _, f := range functions {
		store.Insert(f)
	}
	return store
}

// Insert adds a known function into the store.
// Panics if a named function is inserted.
func (s *FunctionsStore) Insert(function Function) {
	if function.kind != functionKindKnown {
		panic("only known functions can be inserted into FunctionsStore")
	}
	name := function.Name()
	s.dict[function.value] = functionEntry{function: function, name: name}
}

// AssignedName returns the assigned name for a function if it exists in the store.
func (s *FunctionsStore) AssignedName(function Function) (string, bool) {
	if s == nil {
		return "", false
	}
	if function.kind == functionKindKnown {
		entry, ok := s.dict[function.value]
		if !ok {
			return "", false
		}
		return entry.name, true
	}
	return "", false
}

// Name returns the name for a function from the store, falling back to the function's own name.
func (s *FunctionsStore) Name(function Function) string {
	if name, ok := s.AssignedName(function); ok {
		return name
	}
	return function.Name()
}

// NameForFunction returns the name of a function using an optional store.
func NameForFunction(function Function, store *FunctionsStore) string {
	if store != nil {
		if name, ok := store.AssignedName(function); ok {
			return name
		}
	}
	return function.Name()
}

// Clone returns an independent copy of the store.
func (s *FunctionsStore) Clone() *FunctionsStore {
	if s == nil {
		return NewFunctionsStore()
	}
	clone := NewFunctionsStore()
	for _, entry := range s.dict {
		clone.Insert(entry.function)
	}
	return clone
}

// Global functions store singleton.
var (
	globalFunctions     *FunctionsStore
	globalFunctionsOnce sync.Once
)

// GlobalFunctions returns the lazily initialized global functions store.
func GlobalFunctions() *FunctionsStore {
	globalFunctionsOnce.Do(func() {
		globalFunctions = NewFunctionsStore(
			FunctionAdd, FunctionSub, FunctionMul, FunctionDiv,
		)
	})
	return globalFunctions
}
